### DISCUSSION TITLE
Custom elements integration (Web components)

### DISCUSSION THREAD

**Discussion by s-voloshynenko** [nuxt/nuxt#17263] [General]
Hi there, hope it will be useful to someone because I spent a lot of time integrating custom elements (Web Components) in the nuxt3 framework...

The background:
We (company) are going to migrate from Angular + SSR (universal) application to something more performant for customers and efficient from the SSR development point of view. I really like the paradigm/architecture of Nuxt as the development is based on SSR and one of the criteria for migration is the support of custom elements with hydration. Due to performance reasons, in some site areas, we are using Web Components (via Stencil.js compiler) and we can't get rid of them as they do their job very well. In the Nuxt2 framework, it was easy to integrate web components without extra additional configuration and I didn't face build issues (was doing POC with Nuxt2 to provide more data on why we should use Nuxt). You can just import components and define them via [custom elements API](https://developer.mozilla.org/en-US/docs/Web/Web_Components). Something like:
```vue
<template>
  <custom-element/>
</template>

<script>
import { CustomElement } from '<package>/custom-elements';
customElements.define('custom-element', CustomElement);

export default {}
</script>
```
Nuxt3 is based on Vue3 and Vite stack so we should consider Vue3 API to define custom elements and Vite build configuration. Vue3 provides an API to define custom elements via [app.confi.compilerOptions.isCustomElement](https://v3.vuejs.org/guide/web-components.html#using-custom-elements-in-vue). When I try to reach compiler options during SSR (via useNuxtApp().vueApp.config.compilerOptions.isCustomElement) it fails, as it's not accessible during server rendering. I was trying to solve this issue via Vite config using ['@vitejs/plugin-vue'](https://v3.vuejs.org/guide/web-components.html#skipping-component-resolution) but it throws errors I wasn't able to resolve...

Nuxt3 config file provides you an access to Vue compilerOptions:
```ts
export default defineNuxtConfig({
  vue: {
    compilerOptions: {
      isCustomElement: tag => tag.includes('custom-'),
    }
  }
})
```
This configuration resolves issues with Vue so it knows that the element you provided in the template is a custom element. But you still can't just import the custom element within the script, like:
```vue
<script>
import { CustomElement } from '<package>/custom-elements';
customElements.define('custom-element', CustomElement);

export default {}
</script>
```
When you load the script, Nuxt3 application throws errors during SSR as window, HTMLElement and so on are not defined... The only way how you can workaround this:
```vue
<template>
  <custom-element/>
</template>

<script>
export default {
  mounted() {
    if (process.client) {
      import('<package>/custom-elements').then({ CustomElement } => {
        customElements.define('custom-element', CustomElement);
      });
    }
  }
}
</script>
```
So the steps are:
1. Define custom elements in `nuxt.config` file with `vue.compilerOptions.isCustomElement` configuration.
2. Make sure you load the custom element on the client side and define it.

Questions are:
1. Why we can't load custom elements that have access to the window and use HTMLElement interface...
```bash
[Vue Router warn]: uncaught error during route navigation:
ReferenceError: HTMLElement is not defined
```
2. In order to hydrate the custom element, we need to have an access to the HTML output (when server responds to the client). How I can access the final HTML that will be sent to the client? I want to modify it before I send it to the client.

PS: sorry for the detailed background, this is just to highlight that the nuxt team are doing really important things... and will it be useful to add how to define custom elements in nuxt3? Thanks.

**Comment by lucpotage** [comment]
Hi! Did you try that config instead with the `template` property as mentioned in the Vue 3 docs?
https://v3.vuejs.org/guide/web-components.html#skipping-component-resolution

```js
export default defineNuxtConfig({
  vue: {
    template: {
      compilerOptions: {
        isCustomElement: tag => tag.includes('custom-')
      }
    }
  }
})
```

**Comment by thijsw** [comment]
Hi @s-voloshynenko did you happen to find a solution? I just posted more or less the same question on Stackoverflow [1] when I stumbled on this post. I'm very curious if anybody knows a solution to this 'issue'. Thank you!

[1] https://stackoverflow.com/questions/72478699/how-to-use-ssr-with-stencil-in-a-nuxt-3-project

**Reply by GionRubitschung** [reply]
I found a solution for this. I added an answer to your question in stackoverflow: https://stackoverflow.com/a/73036156/10875110

**Reply by thijsw** [reply]
Thank you, I will take a look!

**Comment by s-voloshynenko** [comment]
Solution for this from nuxt@3.3.0

**Update nuxt.config**
```ts
vue: {
  compilerOptions: {
    isCustomElement: (tag) => ['my-', 'etc-'].some(prefix => tag.startsWith(prefix)),
  },
},
```

**Client only**

1. Define client plugin `stencil-definition.client.ts` in `/plugins` folder or via modules.
2. Import components definition, e.g.:

```ts
import { defineCustomElements as defineMyFirstComponent } from '@my-first-component/dist/custom-elements';
import { defineCustomElements as defineMySecondComponent } from '@my-second-component/dist/custom-elements';

export default defineNuxtPlugin(() => {
  defineMyFirstComponent();
  defineMySecondComponent();

  // or you can use dynamic path, e.g. if you have dynamic import based on locale and want to track readiness, etc
  const locale = useLocaleConfig();
  const myThirdComponentState = useThirdComponentState();

  import(`@my-third-component/dist/${locale.value}/custom-elements/my-third-component`).then(() => {
    myThirdComponentState.componentLoaded();
  });
});

```

**Server hydration**

1. Define plugin `stencil-hydration.plugin.ts` in `/server/plugins` folder or via modules.
2. Hydrate components on `render:response` nitro hook:

```ts
import { renderToString } from '@my-first-component/hydrate';

export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:response', async (response) => {
    response.body = (await renderToString(response.body, {
        prettyHtml: false,
        removeBooleanAttributeQuotes: false,
        removeEmptyAttributes: true,
        removeHtmlComments: true,
        removeAttributeQuotes: true,
    })).html;
  });
});
```

**Comment by cesarecaoduro** [comment]
Morning all,
I know this looks like has been close but I was wondering if there is a better solution with newer versions of Nuxt?
I am trying to use a package that provides web-components created using Lit and it looks like this example requires me to almost write a custom importer for each of the component?
Just as reference this is the package https://npm.io/package/@thatopen/ui.
@s-voloshynenko would love your input here if possible, but I would appreciate any help that could come from the group.

**Reply by s-voloshynenko** [reply]
Hi @cesarecaoduro , I didn't work with Nuxt last years but I assume that the approach is still the same according to the documentation. Here is the summary from what I can see:
1. This package https://npm.io/package/@thatopen/ui provides you with Web Components so it uses browsers API to define (register) custom elements which is not available during the server side rendering.
2. Nuxt uses hydration approach which allows you to render components on the server side and then hydrate them during client side rendering. As I can see from the package you provided, it doesn't generate hydration bundle output, so the only way to use those components is to define them on Client Side only.

So the result of this integration should be the following:
1. Make sure you define (register) those components only on client side. E.g. you can define your own plugin with `.client` extension, so it will fire only on browser side. Reference: https://nuxt.com/docs/guide/directory-structure/plugins. In this plugin, you should register custom components from that library, according to docs:
```js
import { UIManager } from "@thatopen/ui"

export default defineNuxtPlugin((nuxtApp) => {
  UIManager.init()
})

```
2. You can use those custom components within `<ClientOnly>` tag, so it will be ignored during server side rendering. Reference: https://nuxt.com/docs/api/components/client-only:
```html
<ClientOnly>
  <bim-text-input label="Tool Name" value="BCFManager"></bim-text-input>
  <bim-input label="Position" vertical>
    <bim-number-input pref="X" min="1" value="10" max="50" suffix="m" slider></bim-number-input>
    <bim-number-input pref="X" min="1" value="20" max="50" suffix="m" slider></bim-number-input>
    <bim-number-input pref="X" min="1" value="30" max="50" suffix="m" slider></bim-number-input>
  </bim-input>
</ClientOnly>
```

As I can see, https://npm.io/package/@thatopen/ui uses Lit, so potentially they could provide you with server components so you will be able to load them on Server side as well https://lit.dev/docs/ssr/server-usage/. But there are no any reference on their's documentation.

The problem with Client only components is that you will see an empty space until client bundle is loaded, hydrated & initialized. This leads to CLS issues and sometimes to performance degration in case you will decide to use those components across the whole app.
